Functional test case

eschwartz opened the pull request:
I put together some basic end-to-end functional tests against the sculpin binary. Each test creates a temporary `SculpinTestProject` directory, to which you can add whatever source files / layouts / config your want, run the sculpin binary, then make sure the generated output is as expected.

These tests use a base `FunctionalTestCase` class, which provides a number of utilities for writing functional tests as I have described.

My motivation for this is:

- I noticed that there is test coverage missing for several of the bundles
- As a contributor, it's much less scary for me to propose changes if I know I'm not breaking existing behavior
- As a contributor, there is a lower barrier to entry to me to write functional tests than unit tests.

Specifically, I'm planning to do some work on the themes functionality, and I'd like to make sure that I'm not breaking what's already there, and that I can make sure any additions I make are well tested.

davedevelopment commented:
:+1: in general.

I think you should tear down the file system in the `setupTestProject` method, I prefer to leave artifacts intact for inspection and also means the file system will definitely be cleaned up before a test runs.

What is the `Bootstrap.php` file for? Is this a sculpin thing?

eschwartz commented:
> I think you should tear down the file system in the setupTestProject method, I prefer to leave artifacts intact for inspection and also means the file system will definitely be cleaned up before a test runs.

[Good call.](https://github.com/eschwartz/sculpin/commit/3ab8eafb724043060859148d16466bccb3469678)

> What is the `Bootstrap.php` file for? Is this a sculpin thing?

PHPUnit can be configured with a bootstrap script, which gets run once before any tests are executed. Often you'll use this to do your autoloading, etc. I added this file trying to debug some issues I was having running tests in my IDE. In the end it wasn't needed though -- [I've deleted the file.](https://github.com/eschwartz/sculpin/commit/4904c0375d4b63c696219a2a5fd4202dee0fdd8c)

davedevelopment commented:
Do we need the `src/Sculpin/Tests` dir adding to the `phpunit.xml.dist`?

davedevelopment commented:
@eschwartz could you add the `src/Sculpin/Tests` dir to the `phpunit.xml.dist` so these tests run with the rest of the suite and we can get some more opinions on this

eschwartz commented:
Yeah, sorry, I kind of forgot about this.

I pushed up the change to `phpunit.xml`. I also rebased the branch onto master, to make it easier to merge.

davedevelopment commented:
@eschwartz not a problem and thanks!

Everyone else, what do we think to this?

simensen commented:
I like the idea! I think it would be good to get this in so we can start adding onto it.

One thing, though, is that travis is failing some of the PHP 5.3 stuff because `[]`'s were used in some places. If we get that fixed I think we can merge.

davedevelopment commented:
@simensen done
